From fe6e59b663fc467b831bf0db9fe379db1615b583 Mon Sep 17 00:00:00 2001
From: Ramanathan Choodamani <quic_rchoodam@quicinc.com>
Date: Wed, 22 Feb 2023 01:08:03 -0800
Subject: [PATCH 1/3] hostapd: Fix support for RU puncturing in 320 MHz

Problem description:
For EHT320 operating AP, channel width sent by
AP is considered incorrect by the STA, when puncturing
is enabled.

This change will fix puncturing support for EHT320
and alter the width to correct value.

Signed-off-by: Ramanathan Choodamani <quic_rchoodam@quicinc.com>
---
 src/ap/ieee802_11.c | 28 +++++++++++++++++++++++++++-
 1 file changed, 27 insertions(+), 1 deletion(-)

diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index 17669d7..99e2c3c 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -8527,6 +8527,27 @@ static void ru_punct_update_bw_160(u8 bitmap, u8 pri_chan, u8 *width, u8 *seg0)
 	}
 }
 
+static void ru_punct_update_bw_320(u16 bitmap, u8 pri_chan, u8 *width,
+		u8 *seg0)
+{
+	if (pri_chan < *seg0) {
+		*seg0 -= 16;
+		if (bitmap & 0x00FF) {
+			*width = 1;
+			ru_punct_update_bw_160(bitmap & 0xFF,
+					       pri_chan, width,
+					       seg0);
+		}
+	} else {
+		*seg0 += 16;
+		if (bitmap & 0xFF00) {
+			*width = 1;
+			ru_punct_update_bw_160((bitmap & 0xFF00) >> 8,
+					       pri_chan, width,
+					       seg0);
+		}
+	}
+}
 
 void ru_punct_update_bw(u16 bitmap, u8 pri_chan, u8 *width, u8 *seg0, u8 *seg1)
 {
@@ -8544,7 +8565,12 @@ void ru_punct_update_bw(u16 bitmap, u8 pri_chan, u8 *width, u8 *seg0, u8 *seg1)
 		ru_punct_update_bw_160(bitmap & 0xFF, pri_chan, width, seg0);
 	}
 
-	/* Add for 320 MHz */
+	/* 320 MHz */
+	if ((*width) == 9 && (bitmap & 0xFFFF)) {
+		*width = 2;
+		ru_punct_update_bw_320(bitmap & 0xFFFF, pri_chan, width, seg0);
+	}
+
 }
 
 #endif /* CONFIG_NATIVE_WINDOWS */
-- 
2.7.4

